Major update

maltelueken opened the pull request:
This PR renames the main function (ProphetLinear -> Prophet) because now it contains both linear and logistic growth. It also adds a new variable field where users can differentiate between data used for model fitting and data used for forecasting (called History Indicator).

Adds:
- Logistic growth
- Proper error handling
- NA handling

Improves:
- Cross validation (timezone = "UTC", which does not include daylight savings and prevents bugs)
- QML interface
- Plot axes

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+
+  options <- .prophetInitOptions(jaspResults, options)
+  
+  ready <- (options$dependent != "" && options$time != "")
```
just a nitpick: in JASP we prefer subsetting lists like `options[["dependent"]]` rather than with a dollar sign. Not crucial but just so you know.

maltelueken replied:
Is there a particular reason for that or is that a style thing? 😃

Kucharssim replied:
Apart from style, I think the most important reason is to avoid partial matching. I am not sure if there are more reasons to it. But as I said, I don't think it's crucial (I am also guilty of not following it 100%). Maybe other people have stronger opinions (@AlexanderLyNL ?)

AlexanderLyNL replied:
Indeed, the problem is partial matching and can result in some obscure errors. Partial matching makes R more flexible, but sometimes too flexible, which is why I discourage it in packages.

maltelueken replied:
Ah, ok, that makes sense. Thanks for the explanation!

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+  prophetModelResults <- jaspResults[["prophetResults"]][["prophetModelResults"]]$object
+  
+  title <- switch(options$estimation,
+                  map  = "Changepoint Estimates Table",
+                  mcmc = "Changepoint Posterior Summary Table")
```
```suggestion
                  map  = gettext("Changepoint Estimates Table"),
                  mcmc = gettext("Changepoint Posterior Summary Table"))
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+      delta = delta
+    ))
+  } else {
+    deltaCri <- apply(prophetModelResults$params$delta, 2, quantile, probs = c(0.025, 0.975))
```
Can we have these probabilities changeable as well?

maltelueken replied:
Yes, we can! I made that for the other tables and the parameter plots too!

AlexanderLyNL replied:
Are "params" and "delta" names of lists? If so, I'd recommend calling them directly/unambiguously via `prophetModelResults[["params"]][["delta"]]`

maltelueken replied:
Yes, I will do that from now on (and change the previous)!

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+  x <- try(as.POSIXct(prophetPredictionResults[["ds"]], tz = "UTC"))
+
+  if (isTryError(x) && options$predictionType == "nonperiodicalPrediction")
+    stop("Please enter valid 'Start date' and 'End date' arguments for nonperiodical prediction (e.g., yyyy-mm-dd)")
```
Shoud this be `gettext`ed as well?

maltelueken replied:
Yes, my bad!

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+
+  prophetEvaluationResults <- jaspResults[["prophetResults"]][["prophetEvaluationResults"]]$object
+  
+  prophetPerformancePlots <- createJaspContainer("Performance Plots")
```
```suggestion
  prophetPerformancePlots <- createJaspContainer(gettext("Performance Plots"))
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+.prophetCreatePerformancePlotMse <- function(prophetPerformancePlots, options, prophetEvaluationResults) {
+  if (!is.null(prophetPerformancePlots[["prophetPerformancePlotMse"]])) return()
+  
+  prophetPerformancePlotMse <- createJaspPlot(title = "Performance Plot MSE", height = 320, width = 480)
```
```suggestion
  prophetPerformancePlotMse <- createJaspPlot(title = gettext("Performance Plot MSE"), height = 320, width = 480)
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+.prophetCreatePerformancePlotRmse <- function(prophetPerformancePlots, options, prophetEvaluationResults) {
+  if (!is.null(prophetPerformancePlots[["prophetPerformancePlotRmse"]])) return()
+  
+  prophetPerformancePlotRmse <- createJaspPlot(title = "Performance Plot RMSE", height = 320, width = 480)
```
```suggestion
  prophetPerformancePlotRmse <- createJaspPlot(title = gettext("Performance Plot RMSE"), height = 320, width = 480)
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+.prophetCreatePerformancePlotMape <- function(prophetPerformancePlots, options, prophetEvaluationResults) {
+  if (!is.null(prophetPerformancePlots[["prophetPerformancePlotMape"]])) return()
+  
+  prophetPerformancePlotMape <- createJaspPlot(title = "Performance Plot Mape", height = 320, width = 480)
```
```suggestion
  prophetPerformancePlotMape <- createJaspPlot(title = gettext("Performance Plot Mape"), height = 320, width = 480)
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+
+    ggplot2::scale_x_continuous(name = gettext("Horizon"), breaks = xBreaks, limits = range(xBreaks)) +
+
+    ggplot2::scale_y_continuous(name = gettext(stringr::str_to_upper(type)), breaks = yBreaks, limits = range(yBreaks))
```
any reason why not just use `toupper()`?

maltelueken replied:
I didn't know the function existed in base R...

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+  
+  prophetModelResults <- jaspResults[["prophetResults"]][["prophetModelResults"]]$object
+  
+  prophetParameterPlots <- createJaspContainer("Parameter Plots")
```
```suggestion
  prophetParameterPlots <- createJaspContainer(gettext("Parameter Plots"))
```

Kucharssim commented on R/prophet.R:
```diff
@@ -0,0 +1,1189 @@
+  prophetParameterPlotMarginal <- createJaspContainer(title = gettext("Posterior Distributions"))
+  
+  parNames <- c("k", "m", "sigma_obs")
+  parTitles <- c("Growth rate", "Offset", "Residual variance")
```
```suggestion
  parTitles <- c(gettext("Growth rate"), gettext("Offset"), gettext("Residual variance"))
```

Kucharssim commented on inst/qml/Prophet.qml:
```diff
@@ -0,0 +1,598 @@
+	VariablesForm
+	{
+		AvailableVariablesList	{ name: "allVariablesList" }
+		AssignedVariablesList	{ name: "dependent";		title: qsTr("Dependent Variable");	suggestedColumns: ["scale"]; singleVariable: true													}
+		AssignedVariablesList	{ name: "time";				title: qsTr("Time");				suggestedColumns: ["nominal"]; singleVariable: true													}
+		AssignedVariablesList	{ name: "changepoints";		title: qsTr("Changepoints");		suggestedColumns: ["nominal"]; singleVariable: true													}
+		AssignedVariablesList	{ name: "capacity";			title: qsTr("Carrying Capacity");	suggestedColumns: ["scale"]; singleVariable: true; id: cap; enabled: growth.value === "logistic"	}
+		AssignedVariablesList	{ name: "minimum";			title: qsTr("Saturating Minimum");	suggestedColumns: ["scale"]; singleVariable: true; id: floor; enabled: growth.value === "logistic"	}
+		AssignedVariablesList	{ name: "covariates";		title: qsTr("Covariates");			suggestedColumns: ["scale"]; id: covs																}
+		AssignedVariablesList	{ name: "historyIndicator";	title: qsTr("History Indicator");	suggestedColumns: ["nominal"]; singleVariable: true													}
```
There is some problem with indenting is seems

maltelueken replied:
Fixed that!

Kucharssim commented on inst/qml/Prophet.qml:
```diff
@@ -0,0 +1,598 @@
+		AssignedVariablesList	{ name: "changepoints";		title: qsTr("Changepoints");		suggestedColumns: ["nominal"]; singleVariable: true													}
+		AssignedVariablesList	{ name: "capacity";			title: qsTr("Carrying Capacity");	suggestedColumns: ["scale"]; singleVariable: true; id: cap; enabled: growth.value === "logistic"	}
+		AssignedVariablesList	{ name: "minimum";			title: qsTr("Saturating Minimum");	suggestedColumns: ["scale"]; singleVariable: true; id: floor; enabled: growth.value === "logistic"	}
+		AssignedVariablesList	{ name: "covariates";		title: qsTr("Covariates");			suggestedColumns: ["scale"]; id: covs																}
```
Maybe you could increase the size of this box for covariates so that it is clear that you can have more of them

maltelueken replied:
Yes, I didn't know that was possible, but it's a good idea!

Kucharssim reviewed (changes requested):
Wow it looks *really* nice now! The code is also very clean and well written, I left just some minor comments.

There is one thing regarding functionality/UX, that is exporting the predictions. I am unable to get the .csv exported. I select the target location, then the analysis runs again, but after it ends the file is not in the location I selected. Am I supposed to do something more or is this something wrong going on there?

maltelueken commented:
Thanks for the feedback! I fixed all the gettext() things, added a custom CI option, and increased the covariate variable form. Also some small bugs. Saving the predictions should work now, too.

I thought it might be nice to have the changepoints in the forecast plots, too, so I added that as well.

The title of the covariate options form looks a bit weird, but apparently there is no better way of doing that currently.

Kucharssim commented on inst/qml/Prophet.qml:
```diff
@@ -0,0 +1,626 @@
+		AssignedVariablesList	{ name: "capacity"; title: qsTr("Carrying Capacity"); suggestedColumns: ["scale"]; singleVariable: true; id: cap; enabled: growth.value === "logistic"		}
+		AssignedVariablesList	{ name: "minimum"; title: qsTr("Saturating Minimum"); suggestedColumns: ["scale"]; singleVariable: true; id: floor; enabled: growth.value === "logistic"	}
+		AssignedVariablesList	{ name: "covariates"; title: qsTr("Covariates"); suggestedColumns: ["scale"];																				}
+		AssignedVariablesList	{ name: "historyIndicator"; title: qsTr("History Indicator"); suggestedColumns: ["scale"]; singleVariable: true									}
```
So I am thinking about this input, as it is quite unusual in JASP. I am not sure whether it will be obvious how to use it (the reason for my worry is that it took me a while to realise what this does). The obvious solution would be to bother @JorisGoosen and propose implementing a `boolean` type column, but I don't know whether it's a good idea. 

At the minimum, we could think if renaming this variable to something that everyone will understand would help. Something like `"Include in training (boolean 1/0)"` or something like that.

Other option would be to see how is similar stuff done in machine learning, maybe you could just have an input on what is the first date that is used for forecasting and not fitting, and then everything after that date is assumed to have `historyIndicator == 0`.

maltelueken replied:
I also think this is not really an optimal solution, yet, but it is the same the Machine Learning module has (but with a variable form instead of a dropdown). Having a cutoff value for predictions would also work, but is less flexible too. But if you think that it is easier to understand, I will change that! Otherwise, I will rename the variable form and maybe make the error messages more comprehesible.

Kucharssim replied:
Let's stick to renaming this input. I saw that some error messages are already there, so that should be fine. Also remember to properly describe the expected input in the help files...

Kucharssim commented on inst/qml/Prophet.qml:
```diff
@@ -0,0 +1,626 @@
+		preferredHeight: 400 * preferencesModel.uiScale
+		AvailableVariablesList	{ name: "allVariablesList" }
+		AssignedVariablesList	{ name: "dependent"; title: qsTr("Dependent Variable");	suggestedColumns: ["scale"]; singleVariable: true													}
+		AssignedVariablesList	{ name: "time"; title: qsTr("Time"); suggestedColumns: ["nominal"]; singleVariable: true																	}
```
You also managed quite well handling this, nice! I think in the future we could also think about implementing a time/date type column, that would help a lot, too. At the moment it still feels a bit wonky and puts a lot of requirements on the user to format the time/date column exactly how you expect. @JorisGoosen do you have an idea how hard/problematic/stupid would be implementing time/date as a type of column in the JASP data viewer?

JorisGoosen replied:
https://github.com/jasp-stats/INTERNAL-jasp/issues/312 already exists.
I do not think it is particularly hard but till now nobody really seemed to need it.

As in, some people complained that the date-time in JASP looks completely messed up when loading from SPSS or so. So I guess they would be happy.

If this would be useful for this module I could see about adding this in the nearfuture, shouldn't be so bad.
But it would good to know what kind of data R then expects. Maybe either of you could add a description of how you/prophet likes to get its dates formatted. That way I can make sure it is delivered in the right format.

maltelueken replied:
Yes, indeed, Prophet can only handle nominal dates/times from JASP when they're in ISO format (which is R's default). That is they need to be in the format "YYYY-mm-dd hh:mm:ss" where the ":ss" or "hh:mm:ss" can be omitted and "-" can be replaced by "/". But I think I could change the conversion so that any other uniquely identifiable format is the default (but it would need to be the only format to convert).

But in general, having a date/time type in JASP would give the module more flexibility!

JorisGoosen replied:
I copied your remarks to that issue, because otherwise I need to keep following links all through github when I get time to work on it ;)

maltelueken replied:
Oh yes, thanks!

Kucharssim reviewed (changes requested):
I think it is fine now. I have some more general comments that perhaps require some discussion outside of the module. 

I think the module would greatly benefit from catching the status of Stan models for displaying progress bar in JASP, because it really can take a while to run the model. I know that @vandenman has enough of more important stuff to do, just mentioning it here so that this issue is not forgotten :)

Kucharssim commented:
Also, I forgot - would be good to add help file as well!

maltelueken commented:
A help file is on the way and unit tests are ready too. I thought I add them in a separate PR, but I can also add them here, if that is better?

Kucharssim commented:
Ok cool! You can add it here at least we will know if already this PR passes the tests

maltelueken commented:
There are quite a few unit tests because everything needs to be tested for every type of date/time input (secs, mins, hours, days, weeks, years). However, it doesn't take very long to run (maybe a minute), because it only computes 10 MCMC samples per test.

JorisGoosen commented:
> I think the module would greatly benefit from catching the status of Stan models for displaying progress bar in JASP, because it really can take a while to run the model. I know that @vandenman has enough of more important stuff to do, just mentioning it here so that this issue is not forgotten :)

Maybe make a real issue?

Kucharssim commented:
> Maybe make a real issue?

The issue already exists somewhere, I was lazy to find it... here it is https://github.com/jasp-stats/INTERNAL-jasp/issues/1192

Kucharssim commented on tests/testthat/test-prophet.R:
```diff
@@ -0,0 +1,663 @@
+dateTimeUnits <- c("years", "weeks", "days", "hours", "mins", "secs")
+
+test_that("Posterior Summary Table results match (linear)", {
+  options <- analysisOptions("Prophet")
```
```suggestion
  options <- jaspTools::analysisOptions("Prophet")
```

Kucharssim commented on tests/testthat/test-prophet.R:
```diff
@@ -0,0 +1,663 @@
+  for (i in 1:6) {
+    options$time <- dateTimeVarNames[i]
+    set.seed(1)
+    results <- runAnalysis(name = "Prophet", dataset = "prophetTest.csv", options = options)
```
```suggestion
    results <- jaspTools::runAnalysis(name = "Prophet", dataset = "prophetTest.csv", options = options)
```

and the same elsewhere

Kucharssim commented on tests/testthat/test-prophet.R:
```diff
@@ -0,0 +1,663 @@
+  options$changePointTable <- TRUE
+  options$predictionSavePath <- ""
+  set.seed(1)
+  results <- runAnalysis("Prophet", "~/JASP/jaspProphet/tests/testthat/prophetTest.csv", options)
```
```suggestion
  results <- runAnalysis("Prophet", "prophetTest.csv", options)
```

Kucharssim commented on tests/testthat/test-prophet.R:
```diff
@@ -0,0 +1,663 @@
+  results <- runAnalysis("Prophet", "prophetTest.csv", options)
+  expect_identical(results[["status"]], "validationError", label = "'Constant Carrying Capacity' must always be larger than 'Constant Saturating Minimum'")
+
+  options <- analysisOptions("Prophet")
+  options$dependent <- "contNormal"
+  options$time <- "dateDay"
+  options$mcmcSamples <- 10
+  options$predictionSavePath <- ""
+  options$seasonalities <- list(list(name = "---",
+                                     period = 4,
+                                     unit = dateTimeUnits[i],
+                                     priorSigma = 10,
+                                     fourierOrder = 3,
+                                     mode = "additive"))
+  set.seed(1)
+  results <- runAnalysis("Prophet", "prophetTest.csv", options)
+  expect_identical(results[["status"]], "validationError", label = "Seasonality names must only contain letters, number, dots, or underscores and must start with letters or dots that are not followed by a number")
```
The test here fails because of `dateTimeUnites[i]` - variable `i` does not exist. Not sure what you want to do with it, but it needs a fix ;)

Kucharssim commented on tests/testthat/test-prophet.R:
```diff
@@ -0,0 +1,663 @@
+  set.seed(1)
+  results <- runAnalysis("Prophet", "prophetTest.csv", options)
+  table <- results[["results"]][["prophetMainContainer"]][["collection"]][["prophetMainContainer_prophetTable"]][["data"]]
+  jaspTools::expect_equal_tables(table,
+    list(-2.50859195109382, -2.31656506358638, 0.32805219661169))
```
These test values fail on my computer, are you sure they are correct?

maltelueken replied:
Hmm, I checked them again and they seem to be correct! The tests succeed on my computer... could it be a Windows/Mac thing again?

Kucharssim reviewed (changes requested):
The tests need some fixes themselves :) 

It would be good to set up GitHub actions on this repo as well. Who is responsible for setting it up, @vandenman? Or am I supposed/allowed to do it?

maltelueken commented:
I fixed all the unit-tests (except the one issue with the values) and made some minor label changes in the QML interface.

I also added a help file!

vandenman commented:
> It would be good to set up GitHub actions on this repo as well. Who is responsible for setting it up, @vandenman? Or am I supposed/allowed to do it?

@Kucharssim I just set it up. You are definitely allowed to do it btw, you only need to copy one file. 

@maltelueken The tests should get triggered automatically after your next push to this PR.

Kucharssim commented:
Thanks @vandenman!

@maltelueken can you push once more? That way we'll see whether the test fails here as well (suggesting it is indeed some Mac quirk), or whether it's just something with the setup on my computer.

maltelueken commented:
Indeed, it seems to be a mac issue. Unfortunately, I can't really check it out. @Kucharssim do you know of any issues with MAP estimation in Stan on Mac?
Could you maybe also describe by how much the numbers differ?

vandenman commented:
@maltelueken if you look at the log files then you'll see that everything fails big time on mac. This is probably unrelated to @Kucharssim 's issue. Some things you could fix are the following:

- [ ] `Error in as.POSIXlt.character(as.character(x), ...) : character string is not in a standard unambiguous format`. Not sure where this comes from.
- [ ] `Warning messages:  In readLines(f, n) : incomplete final line found on 'D:/a/jaspProphet/jaspProphet/DESCRIPTION'`. Add a newline at the end of the DESCRIPTION file (and change your settings in Rstudio it always does that for every file).

You can ignore this warning: `Warning messages: package 'testthat' was built under R version 3.6.3`

I'll have a look at the mac issue, but the mac actions has been a little weird lately...

vandenman commented:
The errors on mac are btw:
```
This analysis terminated unexpectedly.   Error in matrix(m$params[[name]], nrow = n.iteration): 'data' must be of a vector type, was 'NULL'
```
and
```
This analysis terminated unexpectedly.   Error in sampler$call_sampler(c(args, dotlist)): empty_nested() must be true before calling recover_memory()
```
you can see the full stack trace if you look at the log files.

maltelueken commented:
> * [ ]  `Error in as.POSIXlt.character(as.character(x), ...) : character string is not in a standard unambiguous format`. Not sure where this comes from.

This error is quite strange because I never used this function (only as.POSIXct), but maybe it is an issue inside the prophet package?

> * [ ]  `Warning messages:  In readLines(f, n) : incomplete final line found on 'D:/a/jaspProphet/jaspProphet/DESCRIPTION'`. Add a newline at the end of the DESCRIPTION file (and change your settings in Rstudio it always does that for every file).

Sure, I'll do that!

> ```
> This analysis terminated unexpectedly.   Error in matrix(m$params[[name]], nrow = n.iteration): 'data' must be of a vector type, was 'NULL'
> ```

> ```
> This analysis terminated unexpectedly.   Error in sampler$call_sampler(c(args, dotlist)): empty_nested() must be true before calling recover_memory()
> ```

Both these errors seem to come from inside the prophet package as well! In the beginning, we had an issue with Prophet and a specific mac version, could that be related?

Kucharssim commented:
> Both these errors seem to come from inside the prophet package as well! In the beginning, we had an issue with Prophet and a specific mac version, could that be related?

Ugh, yeah, this is probably related to: https://github.com/facebook/prophet/issues/1639 and possibly relates to problems of `rstan` on some MacOS versions. I am not sure how to fix it @vandenman - on my Mac computer it was fixed by installing `prophet` from source, the GitHub issues of `prophet` also suggest some alternative workarounds as well. Any ideas?

vandenman commented:
well if you change this:

https://github.com/maltelueken/jaspProphet/blob/05f470e03b616187180236b377e1d3ea3be2898a/tests/testthat.R#L1-L4

into:

```r
if (Sys.getenv("GITHUB_ACTIONS") == "true" && .Platform$OS.type == "unix") {
  install.packages("prophet", type = "source")
} else { # just to make sure which branch we're going into
  print("not installing prophet from source")
}
library(jaspTools)
library(testthat)

jaspTools::runTestsTravis(module = getwd())
```

then that should install the prophet package from source, but only during github actions and on macOS. That might be a temporary workaround. @maltelueken could you try this?

maltelueken commented:
This didn't seem to solve the problem, unfortunately...

vandenman commented:
it probably didn't install into the correct directory. Could you try:
```r
  install.packages("prophet", type = "source",
                   lib = "/Users/runner/work/jaspProphet/pkgs/Frameworks/R.framework/Versions/3.6/Resources/library")
```

vandenman commented:
Okay, so that didn't work either. Maybe @TimKDJ has any ideas? Otherwise, I would suggest that 1) @maltelueken removes the hacky `install.packages` again, and 2) we ignore the tests on macOS for now. 

The plan is to switch to a different method for installing all dependencies in the near future (without jasp-required-files) and I hope that that will also magically fix this macOS issue. It looks like this issue is a dependency problem of some kind, but debugging on the CI is just painfully slow.

maltelueken commented:
Ok, I removed the `install.packages` again! @vandenman do I need to disable the unit tests for mac somehow or can we simply ignore them failing?

vandenman commented:
Just letting them fail is probably fine. Once @Kucharssim is happy this can be merged.

Kucharssim reviewed (approved):
Looks good to me now!
